src/coord/src/coord.rs

```diff
@@ -5274,3 +5198,94 @@ impl Coordinator {
             self.finalize_dataflow(df, DEFAULT_COMPUTE_INSTANCE_ID);
     }
 }
+
+/// A mechanism to ensure that a sequence of writes and reads proceed correctly through timestamps.
+mod timeline {
+
+    /// A timeline is either currently writing or currently reading.
+    ///
+    /// At each time, writes happen and then reads happen, meaning that writes at a time are
+    /// visible to exactly reads at that time or greater, and no other times.
+    enum TimelineState {
+        /// The timeline is producing collection updates timestamped with the argument.
+        Writing(mz_repr::Timestamp),
+        /// The timeline is observing collections aot the time of the argument.
+        Reading(mz_repr::Timestamp),
+    }
+
+    /// A type that provides write and read timestamps, reads observe exactly their preceding writes..
+    ///
+    /// Specifically, all read timestamps will be greater or equal to all previously reported write timestamps,
+    /// and strictly less than all subsequently emitted write timestamps.
+    pub struct Timeline {
+        state: TimelineState,
+        advance_to: Option<mz_repr::Timestamp>,
+    }
+
+    impl Timeline {
+        /// Create a new timeline, starting at the indicated time.
+        pub fn new(initially: mz_repr::Timestamp) -> Self {
+            Self {
+                state: TimelineState::Writing(initially),
+                advance_to: Some(initially),
+            }
+        }
+
+        /// Acquire a new timestamp for writing.
+        ///
+        /// This timestamp will be strictly greater than all prior values of `self.read_ts()`,
+        /// and less than or equal to all subsequent values of `self.read_ts()`.
+        pub fn write_ts(&mut self) -> mz_repr::Timestamp {
+            match self.state {
+                TimelineState::Writing(ts) => ts,
+                TimelineState::Reading(ts) => {
+                    self.state = TimelineState::Writing(ts + 1);
+                    ts + 1
+                }
+            }
+        }
+        /// Acquire a new timestamp for reading.
+        ///
+        /// This timestamp will be greater or equal to all prior values of `self.write_ts()`,
+        /// and strictly less than all subsequent values of `self.write_ts()`.
+        pub fn read_ts(&mut self) -> mz_repr::Timestamp {
+            match self.state {
+                TimelineState::Reading(ts) => ts,
+                TimelineState::Writing(ts) => {
+                    self.state = TimelineState::Reading(ts);
+                    self.advance_to = Some(ts + 1);
```

Comment:
This call was a bit surprising to me. I was expecting a rule like "whenever `self.state` is set to `TimelineState::Writing`, `self.advance_to` is set to the new write timestamp." But here we are setting `self.advance_to` when setting the state to `TimelineState::Reading`! Is this just an optimization, because when swapping from write -> read state we _know_ that we will never again be able to write at a time less than `ts + 1`?

Reply:
I'll write more! The point is that as soon as we switch to reading `ts`, we need to cease writing `ts`. That's actually what we are saying with `self.advance_to`: "what have I stopped writing".

Reply:
And if not obvious: we *have* to do this now, because those reads are going to stall otherwise.

Reply:
Ahh, thanks, yes, that makes lots of sense! So there is still a risk that you can hold this wrong, yeah? If you flip from writing to reading, and then block on reading, you better have cleared out `advance_to` first, or you'll block forever.